File: Data Structures & Algorithms/products-of-array-discluding-self/submission-4.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] solution = new int[nums.length];

        solution[0] = 1;
        int product = nums[0];

        for (int i = 1; i < nums.length; i++) {
            solution[i] = product;
            product *= nums[i];
        }

        product = nums[nums.length - 1];

        for (int i = nums.length - 2; i >= 0; i--) {
            solution[i] *= product;
            product *= nums[i];
        }

        return solution;
    }
}  

File: Data Structures & Algorithms/combination-target-sum/submission-4.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        def helper(index, current, total):
            if total == target:
                res.append(current.copy())
                return
            if total > target or index >= len(nums):
                return
            current.append(nums[index])
            helper(index, current, total + nums[index])
            current.pop()
            helper(index + 1, current, total)

        helper(0, [], 0)
        return res